dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,464 @@
 package api
 
 import (
-	"k8s.io/apimachinery/pkg/api/resource"
+	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
 )
 
 // DeviceClass is a vendor or admin-provided resource that contains
-// contraint and configuration information. Essentially, it is a re-usable
-// collection of predefined data that device claims may use.
+// device configuration and requirements. It can be referenced in
+// the device requests of a claim to apply these presets.
 // Cluster scoped.
 type DeviceClass struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
-	Spec DeviceClassSpec `json:"spec,omitempty"`
+	// Config defines configuration parameters that apply to each request in a claim using this class.
+	// They are ignored while allocating the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []ConfigurationParameters `json:"config,omitempty"`
+
+	// Requirements must be satisfied by devices. Applies to all devices of
+	// a claim when the claim references the class and only to the devices
+	// in a request when referenced there.
+	//
+	// +optional
+	// +listType=atomic
+	Requirements []Requirement `json:"requirements,omitempty" protobuf:"bytes,4,opt,name=requirements"`
 }
 
-// DeviceClassSpec provides either a selector to find other
-// classes, or the specific constraints defining this class.
-type DeviceClassSpec struct {
+// ConfigurationParameters must have one and only one field set.
+type ConfigurationParameters struct {
+	Vendor *VendorConfigurationParameters `json:"vendor,omitempty" protobuf:"bytes,1,opt,name=vendor"`
+}
 
-	// One of selector and classCriteria must be populated.
+// VendorConfigurationParameters contains configuration parameters for a driver.
+type VendorConfigurationParameters struct {
+	// DriverName is used to determine which kubelet plugin needs
+	// to be passed these configuration parameters.
+	//
+	// An admission webhook provided by the vendor could use this
+	// to decide whether it needs to validate them.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	DriverName string `json:"driverName,omitempty" protobuf:"bytes,1,opt,name=driverName"`
```

Comment:
Is this optional or required?  If I leave this blank, what does it mean?

Reply:
Required, which is implied by "must be a DNS subdomain". The "omitempty" is wrong - fixed.

Does it make sense to add `+required` even if it has no effect?

Reply:
+required does have effect

Reply:
I tried adding in Kubernetes, but the generated code and OpenAPI spec remained the same (except of course that the additional comment got copied around):

```patch
diff --git a/staging/src/k8s.io/api/resource/v1alpha2/types.go b/staging/src/k8s.io/api/resource/v1alpha2/types.go
index 158f9e55f78..866cd1113e3 100644
--- a/staging/src/k8s.io/api/resource/v1alpha2/types.go
+++ b/staging/src/k8s.io/api/resource/v1alpha2/types.go
@@ -729,6 +729,8 @@ type ResourceClassParametersList struct {
 // VendorParameters are opaque parameters for one particular driver.
 type VendorParameters struct {
        // DriverName is the name used by the DRA driver kubelet plugin.
+       //
+       // +required
        DriverName string `json:"driverName" protobuf:"bytes,1,name=driverName"`
 
        // Parameters can be arbitrary setup parameters. They are ignored while
```

I know that some code relies on `+optional`, but I am not sure where `+required` is relevant.